### ISSUE TITLE
Group get_local_range returns zero

### ISSUE THREAD

**Issue by tomdeakin** [illuhad/hipSYCL#252]
Tested with the CUDA backend.
If the kernel below saves the group size inside the `parallel_for_work_group` scope as below, the variable is always zero.
```size_t local_size = g.get_local_range(0);```

Additionally, if I use the work-group size (`dot_wgsize`) inside the kernel and *have not* saved it inside the `submit()` lambda scope, I get a memory error.

https://github.com/UoB-HPC/BabelStream/blob/55bd47b5e657c0e1fa268e95127cb0714ed6a8b8/SYCLStream.cpp#L209

**Comment by illuhad** [comment]
Can you confirm again that you are on CUDA and not by chance on CPU? On CUDA, that function just calls `detail::get_local_size()` which just returns `{hipBlockDim_x}` (or higher dimensional equivalents). It's not immediately clear to me where this could go wrong.. On CPU on the other hand, things work quite differently.

In any case, this sort of code pattern (reduction using hierarchical parallelism) is part of the unit tests, so I'm surprised there are issues with it.

> Additionally, if I use the work-group size (dot_wgsize) inside the kernel and have not saved it inside the submit() lambda scope, I get a memory error.

Hierarchical parallelism with flexible group sizes is not yet officially supported. I don't think there's much left to do to get it working, but it's not yet formally tested. If you don't pass the work group size to `submit()`, the SYCL implementation has to submit the kernel with an implementation defined physical work group size which may not be the same as the one you pass to your `parallel_for_work_item()` call (the logical group size). So, you will need to be careful to make sure you use the right one.

I'm on vacation until monday, so may take some time until I have the time to look at the issue more in depth.

**Comment by tomdeakin** [comment]
> Can you confirm again that you are on CUDA and not by chance on CPU?
Seems to be using the CUDA device:
```$ ./a.out                                                                                                                                                                           
BabelStream                                                                                                                                                                                                 
Version: 3.4                                                                                                                                                                                                
Implementation: SYCL                                                                                                                                                                                        
Running kernels 100 times                                                                                                                                                                                   
Precision: double                                                                                                                                                                                           
Array size: 268.4 MB (=0.3 GB)                                                                                                                                                                              
Total size: 805.3 MB (=0.8 GB)                                                                                                                                                                              
Using SYCL device GeForce RTX 2080 Ti                                                                                                                                                                       
Driver: hipSYCL 0.8.1-prerelease on HIP/CUDA 10010                                                                                                                                                          
Reduction kernel config: 272 groups of size 1024
```

I'm passing the work-group size to `cgh.parallel_for_work_group()`, but not `queue.submit()` as I'm not sure that can be done.


Have a nice vacation!! 🌴

**Comment by illuhad** [comment]
Alright, I did some more in-depth investigation on this, and it turns out this is actually an interesting one :)

The following minimal code reproduces the issue:
```c++
#include <CL/sycl.hpp>

using namespace cl::sycl;

int main(){
  queue q;
  q.submit([&](handler& cgh){
   cgh.parallel_for_work_group<class test>(range<1>(1), range<1>(128), [=](group<1> g){
      size_t local_size = 128;
      g.parallel_for_work_item([&](h_item<1> item){
        printf("%lu\n", local_size);
      });
    });
  });
  q.wait();
}
```
As it turns out, it's not even necessary to query the group size to trigger this issue. Simply assigning something to `local_size` as above is enough. If the right hand side of the assignment is a function call, clang does not even emit the function call to LLVM IR. Simple instructions like assignments are apparently also skipped.

I'm not entirely sure yet what the root cause is, but I know that it is related to `local_size` being allocated in local memory since it is defined in work group scope. If you move the declaration into `parallel_for_work_item()`, everything works as expected. I suspect that it is related to the fact that, when we move that variable to local memory in the clang plugin, this also changes the storage class to static, which might have implications on how clang can assign values (?).

Note that the iteration variable in the for loop of your reduction code
```c++
for (int offset = local_size / 2; offset > 0; offset /= 2){
  g.parallel_for_work_item([&](h_item<1> item){
    size_t li = item.get_local_id(0);
    if (li < offset)
      wg_sum[li] += wg_sum[li + offset];
  });
}
```
is not affected because hipSYCL allocates it in private memory, which technically violates the spec. This can have substantial performance advantages and is fine for the most common code patterns. And it's not like any of the other vendors has a caveat-free implementation of hierarchical parallelism anyway...

The functionality to allocate in local memory has so far mostly been tested for patterns such as
```c++
T scratch[SIZE];
cgh.parallel_for_work_item([&](h_item<1>){
  ...
});
```
where initialization is not necessary. Here's an implementation of a reduction using hierarchical parallel for in hipSYCL that might help you: https://github.com/illuhad/hipSYCL/blob/4aec8c4a4523a14c4f2ed29ca2d0997983af936d/tests/unit_tests.cpp#L116

> I'm passing the work-group size to cgh.parallel_for_work_group(), but not queue.submit() as I'm not sure that can be done.

Sorry, that's what I meant :) I was confused by the formulation of having it saved in the `submit()` lambda scope and thought you were talking about flexible group ranges, i.e. not setting the group size when calling `parallel_for_work_group()`, which the spec allows.

**Comment by tomdeakin** [comment]
Thanks. That's an interesting bug... 

We need some internal discussion in the WG about this I think.

> Here's an implementation of a reduction using hierarchical parallel for in hipSYCL that might help you:

FWIW, the `SIZE` here has to be compile time constant. It's `constexpr` in the unit tests. If I make it sized at runtime it doesn't work at runtime:
```[hipSYCL Error] check_error: Received unknown HIP error 700, throwing memory_allocation_error.
[hipSYCL Error] task_graph: task_done_callback caught async error,  invoking async handler.
```

**Comment by illuhad** [comment]
> We need some internal discussion in the WG about this I think.

Yes, these kinds of issues are why I personally think hierarchical parallelism should be simplified..

> FWIW, the SIZE here has to be compile time constant. It's constexpr in the unit tests. If I make it sized at runtime it doesn't work at runtime:

Yes, of course regular C++ rules still apply, and C++ doesn't have variable length arrays (unlike C) :) Actually, I'm surprised that it only errors at runtime, and not at compile time. Maybe variable length arrays are some kind of extension in clang CUDA.
For runtime-sized arrays, there are always local accessors :)

**Comment by tomdeakin** [comment]
>Yes, of course regular C++ rules still apply, and C++ doesn't have variable length arrays (unlike C) :) Actually, I'm surprised that it only errors at runtime, and not at compile time. Maybe variable length arrays are some kind of extension in clang CUDA.
For runtime-sized arrays, there are always local accessors :)

C99 has VLAs so C++ compilers are probably just piggybacking on that as an extension. It seems to work just fine in Clang and GCC.

```cpp
#include <iostream>

int main(int argc, char *argv[]) {

  int unknown = atoi(argv[1]);
  std::cout << "allocate array of size " << unknown << std::endl;

  int foobar[unknown];
  foobar[0] = 10;
  for (int i = 0; i < unknown; ++i)
    std::cout << foobar[i] << std::endl;

}
```

**Comment by illuhad** [comment]
>  C99 has VLAs so C++ compilers are probably just piggybacking on that as an extension. It seems to work just fine in Clang and GCC.

I was aware that VLA is supported by default in GCC C++, but clang is more picky about it and rejects it at least in some circumstances. Since hipSYCL is based on clang, I did not expect VLA to work.. See also this SYCL stackoverflow question where the code didn't compile in clang/DPC++ because it relied on VLA:
https://stackoverflow.com/questions/61855921/casting-accessors-to-c-pointers-in-kernel-code-esp-int-nelem
